lib/blemulator/internal/simulation_manager.dart

```diff
@@ -23,18 +30,79 @@ class SimulationManager {
   Future<void> _destroyClient() async {}
 
   Future<void> _startDeviceScan() async {
-    _peripherals.forEach((peripheral) {
-      _scanSubscriptions.add(peripheral
-          .onScan(allowDuplicates: true)
-          .listen((scanResult) {
-            print(scanResult);
-            return _bridge.addScanResult(scanResult);
-          }));
+    _peripherals.values.forEach((peripheral) {
+      _scanSubscriptions
+          .add(peripheral.onScan(allowDuplicates: true).listen((scanResult) {
+        print(scanResult);
+        return _bridge.publishScanResult(scanResult);
+      }));
     });
   }
 
   Future<void> _stopDeviceScan() async {
     _scanSubscriptions.forEach((subscription) => subscription.cancel());
     _scanSubscriptions.clear();
   }
+
+  Future<void> _connectToDevice(String identifier) async {
+    if (_peripherals[identifier] != null) {
+      _connectionStateSubscriptions.putIfAbsent(
+          identifier,
+              () => _peripherals[identifier]
+              .connectionStateStream
+              .listen((connectionState) {
+            _bridge.publishConnectionState(
+                _peripherals[identifier], connectionState);
+
+            if (connectionState == PeripheralConnectionState.disconnected) {
+              _connectionStateSubscriptions[identifier].cancel();
+              _connectionStateSubscriptions.remove(identifier);
+            }
+          }));
+
+      bool canConnect = await _peripherals[identifier].onConnectRequest();
+
+      if (canConnect) {
+        _peripherals[identifier].onConnect();
+      } else {
+        return Future.error(<String, dynamic>{
+          "errorCode": BleErrorCode.DeviceAlreadyConnected,
+          "reason": "Peripheral $identifier is already connected",
+        });
+      }
+    } else {
+      return Future.error(<String, dynamic>{
```

Comment:
We can react for this case at the top of this function and then we will not need this external if.